### PR TITLE
Change `SIGPIPE` ui from `#[unix_sigpipe = "..."]` to `-Zon-broken-pipe=...`

### PR DESCRIPTION
**Pull request by Enselic**
In the stabilization [attempt](https://github.com/rust-lang/rust/pull/120832) of `#[unix_sigpipe = "sig_dfl"]`, a concern was [raised ](https://github.com/rust-lang/rust/pull/120832#issuecomment-2007394609) related to using a language attribute for the feature: Long term, we want `fn lang_start()` to be definable by any crate, not just libstd. Having a special language attribute in that case becomes awkward.

So as a first step towards the next stabilization attempt, this PR changes the `#[unix_sigpipe = "..."]` attribute to a compiler flag `-Zon-broken-pipe=...` to remove that concern, since now the language is not "contaminated" by this feature.

Another point was [also raised](https://github.com/rust-lang/rust/pull/120832#issuecomment-1987023484), namely that the ui should not leak **how** it does things, but rather what the **end effect** is. The new flag uses the proposed naming. This is of course something that can be iterated on further before stabilization.

Tracking issue: https://github.com/rust-lang/rust/issues/97889

### CODE REVIEW

**Review by lnicola — APPROVED** [review]

**Comment by lnicola** [comment]
LGTM from the rust-analyzer side.

File: src/doc/unstable-book/src/compiler-flags/on-broken-pipe.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+The tracking issue for this feature is: [#97889]
+
+Note: The ui for this feature was previously an attribute named `#[unix_sigpipe = "..."]`.
+
+[#97889]: https://github.com/rust-lang/rust/issues/97889
+
+---
+
+The `-Zon-broken-pipe=...` compiler flag  can be used to specify how libstd shall setup `SIGPIPE` on Unix platforms before invoking `fn main()`. This flag is ignored on non-Unix targets. There are three variants:
```
Comment (jieyouxu) [inline src/doc/unstable-book/src/compiler-flags/on-broken-pipe.md:13]:
```suggestion
The `-Zon-broken-pipe=...` compiler flag can be used to specify how libstd shall setup `SIGPIPE` on Unix platforms before invoking `fn main()`. This flag is ignored on non-Unix targets. There are three variants:
```

Hmm, I wonder if a warning should be issued on the platforms that the flag doesn't have any effects on. Probably would be more annoying than helpful.

Should the default behavior be mentioned here immediately as well?

**Review by jieyouxu — APPROVED** [review]
The changes look good to me on the compiler side. I'll roll a libs reviewer for the library part of this PR.

**Comment by jieyouxu** [comment]
Rolling a libs reviewer for the libs part of this PR. If the changes also look good from the libs side, feel free to r=me.

r? libs

File: src/tools/miri/src/eval.rs [hunk 2]
```diff
@@ -387,7 +387,7 @@ pub fn create_ecx<'mir, 'tcx: 'mir>(
             let main_ptr = ecx.fn_ptr(FnVal::Instance(entry_instance));
 
             // Always using DEFAULT is okay since we don't support signals in Miri anyway.
-            // (This means we are effectively ignoring `#[unix_sigpipe]`.)
+            // (This means we are effectively ignoring `SIGPIPE`.)
```
Comment (RalfJung) [inline src/tools/miri/src/eval.rs:390]:
```suggestion
            // (This means we are effectively ignoring `-Zon-broken-pipe`.)
```

Comment (Enselic) [reply src/tools/miri/src/eval.rs]: Done.

File: src/doc/unstable-book/src/compiler-flags/on-broken-pipe.md (hunk 1 above)
Comment (Enselic) [reply src/doc/unstable-book/src/compiler-flags/on-broken-pipe.md]: I agree a warning of other platforms probably would be more annoying than helpful.

I also agree the docs can be improved, for example with regards to the default behavior and child processes, so I have made some restructurings and improvements now.

File: src/doc/unstable-book/src/compiler-flags/on-broken-pipe.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+---
+
+
+## Overview
+
+The `-Zon-broken-pipe=...` compiler flag can be used to specify how libstd shall setup `SIGPIPE` on Unix platforms before invoking `fn main()`. This flag is ignored on non-Unix targets. The flag can be used with three different values or be omitted entirely. It affects `SIGPIPE` before `fn main()` and before children get `exec()`'ed:
+
+| Compiler flag              | `SIGPIPE` before `fn main()` | `SIGPIPE` before child `exec()` |
+|----------------------------|------------------------------|---------------------------------|
+| not used                   | `SIG_DFL`                    | `SIG_DFL`                       |
```
Comment (Enselic) [inline src/doc/unstable-book/src/compiler-flags/on-broken-pipe.md:20]: A typo that I will wait with fixing until after the first review round by the libs team:

```suggestion
| not used                   | `SIG_IGN`                    | `SIG_DFL`                       |
```

File: compiler/rustc_passes/src/entry.rs [hunk 4]
```diff
@@ -111,6 +99,7 @@ fn find_item(id: ItemId, ctxt: &mut EntryContext<'_>) {
                 });
             }
         }
+        _ => (),
```
Comment (Amanieu) [inline compiler/rustc_passes/src/entry.rs:102]: I'd leave this as an explicit match on `EntryPointType::None` so we can better handle new enum variants being added later.

**Review by Amanieu — COMMENTED** [review]
r=me for the library part

File: src/doc/unstable-book/src/compiler-flags/on-broken-pipe.md (hunk 3 above)
Comment (Enselic) [reply src/doc/unstable-book/src/compiler-flags/on-broken-pipe.md]: Done.

File: src/bootstrap/src/core/build_steps/tool.rs [hunk 5]
```diff
@@ -483,6 +483,14 @@ impl Step for Rustdoc {
             features.as_slice(),
         );
 
+        // If the rustdoc output is piped to e.g. `head -n1` we want the process
+        // to be killed, rather than having an error bubble up and cause a
+        // panic.
+        // FIXME: Synthetic #[cfg(bootstrap)]. Remove when the bootstrap compiler supports it.
+        if build_compiler.stage > 0 {
+            cargo.rustflag("-Zon-broken-pipe=kill");
+        }
```
Comment (RalfJung) [inline src/bootstrap/src/core/build_steps/tool.rs:492]: Turns out this regressed https://github.com/rust-lang/rust/issues/123177.

**Pull request merged**